codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/compression/RequestCompressionIntegration.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.compression
+
+import software.amazon.smithy.kotlin.codegen.KotlinSettings
+import software.amazon.smithy.kotlin.codegen.core.CodegenContext
+import software.amazon.smithy.kotlin.codegen.core.KotlinWriter
+import software.amazon.smithy.kotlin.codegen.core.RuntimeTypes
+import software.amazon.smithy.kotlin.codegen.core.withBlock
+import software.amazon.smithy.kotlin.codegen.integration.KotlinIntegration
+import software.amazon.smithy.kotlin.codegen.model.getTrait
+import software.amazon.smithy.kotlin.codegen.model.hasTrait
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolMiddleware
+import software.amazon.smithy.kotlin.codegen.rendering.util.ConfigProperty
+import software.amazon.smithy.kotlin.codegen.rendering.util.ConfigPropertyType
+import software.amazon.smithy.kotlin.codegen.rendering.util.nestedBuilder
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.traits.RequestCompressionTrait
+
+/**
+ * Adds request compression settings if request compression trait is present.
+ * Adds middleware that registers interceptor and provides functionality for request compression if request compression
+ * is not disabled via settings.
+ */
+class RequestCompressionIntegration : KotlinIntegration {
+
+    override fun enabledForService(model: Model, settings: KotlinSettings): Boolean =
+        model.isTraitApplied(RequestCompressionTrait::class.java)
+
+    override fun customizeMiddleware(
+        ctx: ProtocolGenerator.GenerationContext,
+        resolved: List<ProtocolMiddleware>,
+    ): List<ProtocolMiddleware> = resolved + requestCompressionTraitMiddleware
+
+    override fun additionalServiceConfigProps(ctx: CodegenContext): List<ConfigProperty> =
+        listOf(
+            ConfigProperty {
+                name = "requestCompression"
+                symbol = RuntimeTypes.HttpClient.Config.RequestCompressionConfig
+                builderSymbol = RuntimeTypes.HttpClient.Config.RequestCompressionConfig.nestedBuilder
+                toBuilderExpression = ".toBuilderApplicator()"
+                baseClass = RuntimeTypes.HttpClient.Config.CompressionClientConfig
+                propertyType = ConfigPropertyType.Custom(
+                    render = { prop, writer ->
+                        writer.write(
+                            "override val #1L: #2T = builder.#1L?.let { #2T(builder.#1L!!) } ?: #2T{}",
```

Comment:
question: is the `#1L!!` necessary? I would think since you're in the `#1L?.let { ... }` block, the compiler would know it's non-null

Reply:
It would be better to use `it` inside the `let` block anyway.

Reply:
I think I added this because the compiler was giving me an error, which it shouldn't. I'll try it with `it` though

Reply:
Yeah, the compiler would recognize that the builder property is a `var` and thus may change between when it's known to be not null from `?.let` and when it's used again.